MantidQt/MantidWidgets/src/MantidTreeWidget.cpp

```diff
@@ -189,24 +189,33 @@ QList<MatrixWorkspace_const_sptr>
 MantidTreeWidget::getSelectedMatrixWorkspaces() const {
   // Check for any selected WorkspaceGroup names and replace with the names of
   // their children.
-  QSet<QString> selectedWsNames;
+  // We preserve the order, but use a set to avoid adding duplicate workspaces.
+  std::set<QString> selectedWsNameSet;
+  std::vector<QString> selectedWsNameList;
   foreach (const QString wsName, this->getSelectedWorkspaceNames()) {
     const auto groupWs = boost::dynamic_pointer_cast<const WorkspaceGroup>(
         m_ads.retrieve(wsName.toStdString()));
     if (groupWs) {
       const auto childWsNames = groupWs->getNames();
       for (auto childWsName : childWsNames) {
-        selectedWsNames.insert(QString::fromStdString(childWsName));
+        if (selectedWsNameSet.find(QString::fromStdString(childWsName)) ==
```

Comment:
We should iterate over the names via `const &` then we avoid copying the whole container.